### PR TITLE
refactor(dsp-operator): refactor disk pool operator

### PR DESCRIPTION
**Pull request by abhilashshetty04**
### mark dsp as terminating before deleting it

Currently when dsp cr is deleted, cmd gets hung if for some reason pool deletion is failing. Pool status stays as Online evnthough operator is trying to delete pool as part of reconcilation.

Signed-off-by: Abhilash Shetty [abhilash.shetty@datacore.com](mailto:abhilash.shetty@datacore.com)

### add respective control plane pool status in dsp cr
adding control plane pool status in CR. This is as is representation of respective control plane Pool resource state.

Signed-off-by: Abhilash Shetty [abhilash.shetty@datacore.com](mailto:abhilash.shetty@datacore.com)

### added new cr states to align with control plane states
added new cr state field. Reconciling on new states.

Signed-off-by: Abhilash Shetty <abhilash.shetty@datacore.com>

### CODE REVIEW

File: k8s/operators/src/diskpool/crd.rs [hunk 1]
```diff
@@ -118,6 +120,21 @@ impl DiskPoolStatus {
             available: 0,
         }
     }
+
```
Comment (Abhinandan-Purkait) [inline k8s/operators/src/diskpool/crd.rs:139]: doc comment? Can it be pub(crate) ?

**Review by Abhinandan-Purkait — APPROVED** [review]

File: k8s/operators/src/diskpool/crd.rs [hunk 2]
```diff
@@ -56,6 +56,8 @@ pub enum PoolState {
     /// The resource is present, and the pool has been created. The schema MUST
     /// have a status and spec field.
     Online,
+    /// This state is set when we receive delete event on the dsp cr.
+    Terminating,
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs]: I thought we discussed yesterday not doing this and instead separating into 2 different states? Am I misremembering @blaisedias ?

Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: @tiagolobocastro, Now i have added one more field i.e status. This will represent control plane status for the pool.

File: k8s/operators/src/diskpool/crd.rs (hunk 1 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs:139]: Added doc. WIll check on (crate)

File: k8s/operators/src/diskpool/crd.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 use kube::CustomResource;
-use openapi::models::Pool;
+use openapi::models::{pool_status::PoolStatus as RestPoolState, Pool};
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs]: This is a bit confusing, is it status or state?

File: k8s/operators/src/diskpool/crd.rs [hunk 4]
```diff
@@ -66,11 +69,26 @@ pub enum PoolState {
     Error,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, JsonSchema)]
+#[non_exhaustive]
+pub enum PoolStatus {
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs:89]: missing doc comment for PoolStatus type.

File: k8s/operators/src/diskpool/crd.rs [hunk 5]
```diff
@@ -66,11 +69,26 @@ pub enum PoolState {
...
+    /// State is Unknown
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs]: doc comments should be `C`apitalized and end with a `.`

File: k8s/operators/src/diskpool/crd.rs [hunk 6]
```diff
@@ -66,11 +69,26 @@ pub enum PoolState {
...
+    /// the pool is completely inaccessible
+    Faulted,
+}
+
 #[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, JsonSchema)]
 /// Status of the pool which is driven and changed by the controller loop
 pub struct DiskPoolStatus {
     /// The state of the pool
     pub state: PoolState,
+    /// Pool state from respective control plane object.
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs]:
```suggestion
    /// Pool status from respective control plane object.
```

File: k8s/operators/src/diskpool/main.rs [hunk 7]
```diff
@@ -647,7 +672,26 @@ async fn ensure_crd(k8s: Client) {
...
+        info!(
+            "Creating CRD: {}",
+            serde_json::to_string_pretty(&crd).unwrap()
+        );
+
+        let pp = PostParams::default();
+        match dsp.create(&pp, &crd).await {
+            Ok(o) => {
+                info!(crd = ?o.name_any(), "created");
+                // let the CRD settle this purely to avoid errors messages in the console
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: this piece of code is duplicated, please reuse.

File: k8s/operators/src/diskpool/main.rs [hunk 8]
```diff
@@ -672,14 +716,14 @@ fn error_policy(error: &Error, _ctx: Arc<OperatorContext>) -> Action {
         when.to_rfc2822(),
         duration.as_secs()
     );
+    info!("requeue after :{:?}", duration);
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: we already have a message above, why is this needed?

File: k8s/operators/src/diskpool/main.rs [hunk 9]
```diff
@@ -619,20 +645,19 @@ impl ResourceContext {
...
 
-    // the CRD has not been installed yet, to avoid overwriting (and create upgrade issues) only
-    // install it when there is no crd with the given name
-    if crds.iter().count() == 0 {
-        let crd = DiskPool::crd();
+    // Check if there is an existing dsp crd. If yes, Replace it with new generated one.
+    // Create new if it doesnt exist.
+    let mut crd = DiskPool::crd();
+    if let Ok(exiting) = dsp.get("diskpools.openebs.io").await {
+        crd.metadata.resource_version = exiting.resource_version();
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: Should we only patch if version differs?

File: k8s/operators/src/diskpool/main.rs [hunk 10]
```diff
@@ -619,20 +645,19 @@ impl ResourceContext {
...
         info!(
             "Creating CRD: {}",
             serde_json::to_string_pretty(&crd).unwrap()
         );
 
         let pp = PostParams::default();
-        match dsp.create(&pp, &crd).await {
+        match dsp.replace("diskpools.openebs.io", &pp, &crd).await {
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: below says created, but instead should it be `replaced`?

File: k8s/operators/src/diskpool/main.rs [hunk 11]
```diff
@@ -619,20 +645,19 @@ impl ResourceContext {
...
-    let crds = dsp.list(&lp).await.expect("failed to list CRDS");
 
-    // the CRD has not been installed yet, to avoid overwriting (and create upgrade issues) only
-    // install it when there is no crd with the given name
-    if crds.iter().count() == 0 {
-        let crd = DiskPool::crd();
+    // Check if there is an existing dsp crd. If yes, Replace it with new generated one.
+    // Create new if it doesnt exist.
+    let mut crd = DiskPool::crd();
+    if let Ok(exiting) = dsp.get("diskpools.openebs.io").await {
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: I know it was already hardcoded like so, but can we retrieve this value from somewhere instead, eg: `CRD` type?

File: k8s/operators/src/diskpool/main.rs (hunk 8 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Will remove it . It was logger i added for testing

File: k8s/operators/src/diskpool/main.rs (hunk 9 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: resource_version is not present in auto generated crd. it is added by k8s once object is updated. How would we know if version is different?

Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/main.rs]: Not this one, we should bump the CRD api version from v1alpha1 to something higher. Then we can check this, and if != then update.

Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Could be tedious to update crd resource version manually for any changes in crd. If we use this validation we have to bump it everytime right?  With current code it just replaces with crd from the operator. If someone doesnt update crd version new changes will not be patched and new field will not reflect in CR. So thought of just replacing it regardless. Should not impact if there is no change in crd.

File: k8s/operators/src/diskpool/crd.rs (hunk 4 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs:89]: Added

File: k8s/operators/src/diskpool/crd.rs (hunk 5 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: Done

File: k8s/operators/src/diskpool/crd.rs (hunk 6 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: Done

File: k8s/operators/src/diskpool/main.rs (hunk 11 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: now getting crd name from the crd instance that gets created on line 651

File: k8s/operators/src/diskpool/main.rs (hunk 9 above)
Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/main.rs]: If it's tedious then we automate it, but we should update the version, otherwise what's the point of having a version in the CRD?
I'm fine with updating everytime then, but let's update the api version anyway.

File: k8s/operators/src/diskpool/main.rs (hunk 10 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: We create if crd is not there in the cluster. If its there we replace with crd from the dsp operator

File: k8s/operators/src/diskpool/main.rs (hunk 9 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Okay will add it.

File: k8s/operators/src/diskpool/crd.rs [hunk 12]
```diff
@@ -19,6 +19,7 @@ derive = "Default",
 shortname = "dsp",
 printcolumn = r#"{ "name":"node", "type":"string", "description":"node the pool is on", "jsonPath":".spec.node"}"#,
 printcolumn = r#"{ "name":"status", "type":"string", "description":"pool status", "jsonPath":".status.state"}"#,
+printcolumn = r#"{ "name":"poolstatus", "type":"string", "description":"control plane pool status", "jsonPath":".status.status"}"#,
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs]: what should the control-plane status be named as? @blaisedias @avishnu @Abhinandan-Purkait

**Comment by abhilashshetty04** [comment]
Any suggesting on what should be control plane Pool status on dsp cr be called?

File: k8s/operators/src/diskpool/crd.rs [hunk 13]
```diff
@@ -67,22 +71,39 @@ pub enum PoolState {
 }
 
 #[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, JsonSchema)]
-/// Status of the pool which is driven and changed by the controller loop
+#[non_exhaustive]
+/// PoolStatus is Control plane status of a given DSP CR.
+pub enum PoolStatus {
```
Comment (chriswldenyer) [inline k8s/operators/src/diskpool/crd.rs:89]: PoolStatus sounds a lot like PoolState. IMHO it needs a more descriptive name to distinguish it from PoolState, so that it is obvious to developers what it means.

File: k8s/operators/src/diskpool/crd.rs [hunk 14]
```diff
@@ -67,22 +71,39 @@ pub enum PoolState {
...
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, JsonSchema)]
+/// Status of the pool which is driven and changed by the controller loop.
 pub struct DiskPoolStatus {
-    /// The state of the pool
+    /// The state of the pool.
     pub state: PoolState,
-    /// Capacity as number of bytes
+    /// Pool status from respective control plane object.
+    pub status: Option<PoolStatus>,
```
Comment (chriswldenyer) [inline k8s/operators/src/diskpool/crd.rs]: See my earlier comment, state and status mean the same thing, so a more descriptive name for this variable will make it easier for developers.

File: k8s/operators/src/diskpool/crd.rs (hunk 12 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/diskpool/crd.rs]: This can a bit confusing for user, why should a user be interested in the CR status? Can we not display only the control-plane status by default i.e for `kubectl get dsp -n mayastor` and show the CR status if the user does a `kubectl get dsp -owide -n mayastor`. For the naming convention, we can call the control-plane status as the `status` and the existing `status` as `CR Status` which will be only shown on `-owide`.

File: k8s/operators/src/diskpool/main.rs [hunk 15]
```diff
@@ -660,7 +692,7 @@ fn error_policy(error: &Error, _ctx: Arc<OperatorContext>) -> Action {
         Error::ReconcileError { .. } => {
             return Action::await_change();
         }
-        _ => 5,
+        _ => 20,
```
Comment (chriswldenyer) [inline k8s/operators/src/diskpool/main.rs]: It is better practice to use a constant here IMHO.

File: k8s/operators/src/diskpool/crd.rs (hunk 14 above)
Comment (Abhinandan-Purkait) [reply k8s/operators/src/diskpool/crd.rs]: Yeah, the variable names can be `control_plane_status` and `cr_status` respectively for clarity. I agree.

File: k8s/operators/src/diskpool/main.rs [hunk 16]
```diff
@@ -366,6 +351,15 @@ impl ResourceContext {
                 // 'Created' to reflect this.
             }
             Err(e) => {
+                if self.num_retries == 10 {
```
Comment (chriswldenyer) [inline k8s/operators/src/diskpool/main.rs]: It looks like the limit is now hardcoded to 10 instead of being configurable. Why was this done?

File: k8s/operators/src/diskpool/main.rs [hunk 17]
```diff
@@ -619,20 +632,20 @@ impl ResourceContext {
...
 
-    // the CRD has not been installed yet, to avoid overwriting (and create upgrade issues) only
-    // install it when there is no crd with the given name
-    if crds.iter().count() == 0 {
-        let crd = DiskPool::crd();
+    // Check if there is an existing dsp crd. If yes, Replace it with new generated one.
+    // Create new if it doesnt exist.
+    let mut crd = DiskPool::crd();
+    let crd_name = crd.metadata.name.as_ref().unwrap().as_str();
+    if let Ok(exiting) = dsp.get(crd_name).await {
```
Comment (chriswldenyer) [inline k8s/operators/src/diskpool/main.rs]: typo "existing" ?

File: k8s/operators/src/diskpool/main.rs [hunk 18]
```diff
@@ -619,20 +632,20 @@ impl ResourceContext {
...
+        crd.metadata.resource_version = exiting.resource_version();
         info!(
             "Creating CRD: {}",
```
Comment (chriswldenyer) [inline k8s/operators/src/diskpool/main.rs]: It looks like this message should be changed to "Replacing CRD" or something like that?

File: k8s/operators/src/diskpool/crd.rs (hunk 14 above)
Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/crd.rs]: What about `cr_status` and `pool_status`?

Comment (Abhinandan-Purkait) [reply k8s/operators/src/diskpool/crd.rs]: Sounds good.

File: k8s/operators/src/diskpool/main.rs (hunk 16 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: When the 10th retry results in failure we add failure event on the CR. Hence hardcoded it. This doesnt need to be a user configurable paramater IMO. Do you think this should be constant?

File: k8s/operators/src/diskpool/crd.rs (hunk 14 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: I agree. We need to make names more clear.

File: k8s/operators/src/diskpool/crd.rs (hunk 13 above)
Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/crd.rs:89]: This is mirroring the control-plane where State is composed of a Status and other fields, eg: `VolumeState { status: Online, replicas: 3, .... }`
But it becomes messy here with the PoolState above which I think we might be able to rename to `CrStatus` instead?

File: k8s/operators/src/diskpool/main.rs (hunk 16 above)
Comment (chriswldenyer) [reply k8s/operators/src/diskpool/main.rs]: Ok, sounds reasonable for it to be hardcoded.
Having a constant to define it would be prefereable, I think.

File: k8s/operators/src/diskpool/crd.rs (hunk 13 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs:89]: Im okay with CrStatus.

File: k8s/operators/src/diskpool/main.rs (hunk 17 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Will correct

File: k8s/operators/src/diskpool/main.rs (hunk 16 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Removed this hardcoding

File: k8s/operators/src/diskpool/crd.rs [hunk 19]
```diff
@@ -18,7 +18,8 @@ derive = "PartialEq",
 derive = "Default",
 shortname = "dsp",
 printcolumn = r#"{ "name":"node", "type":"string", "description":"node the pool is on", "jsonPath":".spec.node"}"#,
-printcolumn = r#"{ "name":"status", "type":"string", "description":"pool status", "jsonPath":".status.state"}"#,
+printcolumn = r#"{ "name":"state", "type":"string", "description":"pool status", "jsonPath":".status.state"}"#,
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/crd.rs]: description here should be pool state?

File: k8s/operators/src/diskpool/main.rs [hunk 20]
```diff
@@ -239,55 +245,41 @@ impl ResourceContext {
...
-    /// error state will not be deleted.
-    async fn mark_error(&self) -> Result<Action, Error> {
-        let _ = self.patch_status(DiskPoolStatus::error()).await?;
+    /// Mark Pool state as None as couldnt find already provisioned pool in control plane.
+    async fn mark_pool_not_found(&self) -> Result<Action, Error> {
+        self.patch_status(DiskPoolStatus::not_found()).await?;
 
-        error!(name = ?self.name_any(),"status set to error");
-        Ok(Action::await_change())
+        error!(name = ?self.name_any(),"status pool state to blank");
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]:
```suggestion
        error!(name = ?self.name_any(), "Pool not found, clearing status");
```

File: k8s/operators/src/diskpool/main.rs [hunk 21]
```diff
@@ -497,7 +497,7 @@ impl ResourceContext {
             }
             Err(clients::tower::Error::Request(_)) => {
                 // Probably grpc server is not yet up
-                return self.mark_unknown().await;
+                return Ok(Action::requeue(Duration::from_secs(3)))
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: Shouldn't we be marking as unknown here?

File: k8s/operators/src/diskpool/main.rs [hunk 22]
```diff
@@ -479,11 +479,11 @@ impl ResourceContext {
 
                         // We expected the control plane to have a spec for this pool. It didn't so
                         // set the CRD to the error state and don't try to recreate it.
-                        self.mark_error().await
+                        self.mark_pool_not_found().await
                     }
                 } else if response.status() == clients::tower::StatusCode::SERVICE_UNAVAILABLE || response.status() == clients::tower::StatusCode::REQUEST_TIMEOUT {
                     // Probably grpc server is not yet up
-                    self.mark_unknown().await
+                    Ok(Action::requeue(Duration::from_secs(5)))
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: Shouldn't we be marking as unknown here?

File: k8s/operators/src/diskpool/main.rs [hunk 23]
```diff
@@ -600,13 +601,25 @@ impl ResourceContext {
             |event| async move {
                 match event {
                     finalizer::Event::Apply(dsp) => Self::put_finalizer(dsp).await,
-                    finalizer::Event::Cleanup(_dsp) => Self::delete_finalizer(self.clone()).await,
+                    finalizer::Event::Cleanup(dsp) => {
+                        if dsp.status.as_ref().unwrap().cr_state != CrPoolState::Terminating {
+                            info!("set state as terminating");
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: do we need this info here and below?

File: k8s/operators/src/diskpool/main.rs [hunk 24]
```diff
@@ -619,13 +632,34 @@ impl ResourceContext {
...
+                // let the CRD settle this purely to avoid errors messages in the console
+                // that are harmless but can cause some confusion maybe.
+                tokio::time::sleep(Duration::from_secs(5)).await;
+            }
+
+            Err(e) => {
+                error!("failed to create CRD error {}", e);
+                tokio::time::sleep(Duration::from_secs(1)).await;
+                std::process::exit(1);
+            }
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: seems this code is duplicated below, can we reuse?

**Review by tiagolobocastro — COMMENTED** [review]
You will have to rebase.
But before that, do we want to raise it here first or in the release/2.0?
@Abhinandan-Purkait  @blaisedias @avishnu

File: k8s/operators/src/diskpool/crd.rs (hunk 12 above)
Comment (blaisedias) [reply k8s/operators/src/diskpool/crd.rs]: I can see that the new CR status is named as `cr_state` and what I assume is control-plane pool state is `state` I think to disambiguate it should not be named `state`
Maybe `pool_state`  or  pedantically `control_plane_pool_state`.

File: k8s/operators/src/diskpool/main.rs (hunk 22 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: I think so . Should we do it at first ServiceUnavailable itself? Thinking of marking it as None. Control plane status is Option anyway.

File: k8s/operators/src/diskpool/main.rs (hunk 23 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: No needed will remove

File: k8s/operators/src/diskpool/crd.rs (hunk 19 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: Yes

File: k8s/operators/src/diskpool/main.rs (hunk 22 above)
Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/main.rs]: Yes I think so, because we cannot determine what the status is, should be None (or unknown ?).

Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: we dont know so maybe lets not post anything. These erros are connectivity/gRPC issues. So dont know what is the status .

File: k8s/operators/src/diskpool/crd.rs (hunk 14 above)
Comment (chriswldenyer) [reply k8s/operators/src/diskpool/crd.rs]: I thought the suggestion was pool_status?
Using "state " and "status" to distinguish between different things arbitrarily isn't good IMHO.

**Review by chriswldenyer — COMMENTED** [review]
Please use a different name for the pool status.

File: k8s/operators/src/diskpool/crd.rs (hunk 12 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: state is old cr states. status was control plane status. Made it pool_status now for more clarity.

File: k8s/operators/src/diskpool/crd.rs (hunk 14 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: Yes, made it pool_status. This is control plane status for the dsp.

File: k8s/operators/src/diskpool/main.rs (hunk 15 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Done

File: k8s/operators/src/diskpool/main.rs (hunk 22 above)
Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/main.rs]: The problem is that if we don't change state, and the rest/grpc is not back up for whatever reason, we'll keep saying a pool is online for example, when it might not be.
I don't know what the common practice is in k8s for this, anyone knows? @Abhinandan-Purkait  @niladrih

Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: I was thinking of not showing it at all. marking as None. We dont know will not show

File: k8s/operators/src/diskpool/main.rs (hunk 23 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Removed

File: k8s/operators/src/diskpool/crd.rs (hunk 19 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/crd.rs]: made it control plane pool state

File: k8s/operators/src/diskpool/main.rs (hunk 20 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Done

File: k8s/operators/src/diskpool/main.rs (hunk 21 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Setting as None

File: k8s/operators/src/diskpool/main.rs (hunk 22 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Setting as None

File: k8s/operators/src/diskpool/main.rs [hunk 25]
```diff
@@ -838,12 +814,6 @@ async fn main() -> anyhow::Result<()> {
                 .env("JAEGER_ENDPOINT")
                 .help("enable open telemetry and forward to jaeger"),
         )
-        .arg(
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: Don't remove this flag from CLI as it would be breaking change, instead simply not use it.

File: k8s/operators/src/diskpool/main.rs (hunk 24 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Incorporated

File: k8s/operators/src/diskpool/main.rs (hunk 7 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Incorporated

File: k8s/operators/src/diskpool/main.rs (hunk 25 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Alright

Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Done

File: k8s/operators/src/diskpool/main.rs [hunk 26]
```diff
@@ -362,13 +300,75 @@ impl ResourceContext {
...
+                            Ok(Action::requeue(Duration::from_secs(self.ctx.interval)))
+                        }
+                    }
+                    Err(clients::tower::Error::Response(response))
+                        if response.status() == StatusCode::NOT_FOUND =>
+                    {
+                        self.k8s_notify(
+                            "Create or Import Failure",
+                            "Failure",
+                            format!("Unable to find Io engine node {}", &self.spec.node()).as_str(),
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: s/Io engine/io-engine

File: k8s/operators/src/diskpool/main.rs [hunk 27]
```diff
@@ -362,13 +300,75 @@ impl ResourceContext {
...
+                            "Critical",
+                        )
+                        .await;
+                        error!("Unable to find Io engine node {}", &self.spec.node());
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: s/Io engine/io-engine

File: k8s/operators/src/diskpool/main.rs [hunk 28]
```diff
@@ -362,13 +300,75 @@ impl ResourceContext {
...
+                        } else {
+                            self.k8s_notify(
+                                "Create or Import Failure",
+                                "Failure",
+                                format!("Unable to create or import pool {}", error).as_str(),
+                                "Critical",
+                            )
+                            .await;
+                            error!("Unable to create or import pool {}", error);
+                            Ok(Action::requeue(Duration::from_secs(self.ctx.interval)))
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: return error

File: k8s/operators/src/diskpool/main.rs [hunk 29]
```diff
@@ -362,13 +300,75 @@ impl ResourceContext {
...
+                    {
+                        self.k8s_notify(
+                            "Create or Import Failure",
+                            "Failure",
+                            format!("Unable to find Io engine node {}", &self.spec.node()).as_str(),
+                            "Critical",
+                        )
+                        .await;
+                        error!("Unable to find Io engine node {}", &self.spec.node());
+                        Ok(Action::requeue(Duration::from_secs(self.ctx.interval)))
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: return error

File: k8s/operators/src/diskpool/main.rs [hunk 30]
```diff
@@ -362,13 +300,75 @@ impl ResourceContext {
...
+                    _ => {
+                        self.k8s_notify(
+                            "Create or Import Failure",
+                            "Failure",
+                            format!("Unable to create or import pool {}", error).as_str(),
+                            "Critical",
+                        )
+                        .await;
+                        error!("Unable to create or import pool {}", error);
+                        Ok(Action::requeue(Duration::from_secs(self.ctx.interval)))
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: return error

File: k8s/operators/src/diskpool/main.rs [hunk 31]
```diff
@@ -470,20 +451,28 @@ impl ResourceContext {
...
-                    self.mark_unknown().await
+                    self.k8s_notify(
+                        "Unreachable",
+                        "Check",
+                        "Could not reach Rest API service. Please check control plane health",
+                        "Warning",
+                    )
+                        .await;
+                    self.mark_pool_not_found().await?;
+                    Ok(Action::requeue(Duration::from_secs(10)))
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]:
```suggestion
                    self.mark_pool_not_found().await
```

File: k8s/operators/src/diskpool/main.rs [hunk 32]
```diff
@@ -559,36 +541,44 @@ impl ResourceContext {
...
+                            std::env::var("MY_POD_NAME")
+                                .ok()
+                                .unwrap_or_else(|| WHO_AM_I_SHORT.into()),
+                        ),
+                        message: Some(message.into()),
+                        ..Default::default()
+                    },
+                )
+                .await
+                .map_err(|e| error!(?e));
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]:
```suggestion
                .map_err(|error| error!(?error));
```

File: k8s/operators/src/diskpool/main.rs [hunk 33]
```diff
@@ -600,13 +590,32 @@ impl ResourceContext {
...
+                        match self
+                            .pools_api()
+                            .get_node_pool(&self.spec.node(), &self.name_any())
+                            .await
+                        {
+                            Ok(pool) => {
+                                if dsp.status.as_ref().unwrap().cr_state == CrPoolState::Created {
+                                    let new_status = DiskPoolStatus::terminating(pool.into_body());
+                                    let _ = self.patch_status(new_status).await?;
+                                }
```
Comment (tiagolobocastro) [inline k8s/operators/src/diskpool/main.rs]: Why do we check for Create? Otherwise if Creating then we don't transition to Terminating?

**Review by tiagolobocastro — COMMENTED** [review]
Still needs rebasing @abhilashshetty04

File: k8s/operators/src/diskpool/main.rs (hunk 33 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: We want to show Pool usage when marking Terminating. Since we couldnt get pool from the control plane we just delete finalizer for that cr and remove from inventory. We dont need to transition to Terminating for this IMO. There should be nothing holding off on directly deleting cr? Still, we do check if grpc status is NotFound before deletingcr. Or else requeue

Comment (tiagolobocastro) [reply k8s/operators/src/diskpool/main.rs]: Here we could get pool from control-plane, it's just the `CR` that is `Creating`?
What if the pool has replicas? Or if the node is offline? Then we cannot delete the pool, so meanwhile we should show the `cr_state` as `Terminating`, even if the if the previous state was `Creating` and not `Created`? Otherwise it will stay as `Creating` even after user issued delete, which doesn't seem right to me.

@blaisedias is this covered by the BDD, and if so what is the expected behaviour?

File: k8s/operators/src/diskpool/main.rs (hunk 30 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: Done

File: k8s/operators/src/diskpool/main.rs (hunk 31 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: done

File: k8s/operators/src/diskpool/main.rs (hunk 32 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: done

File: k8s/operators/src/diskpool/main.rs (hunk 29 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: done

File: k8s/operators/src/diskpool/main.rs (hunk 28 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: done

File: k8s/operators/src/diskpool/main.rs (hunk 27 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: done

File: k8s/operators/src/diskpool/main.rs (hunk 26 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: done

File: k8s/operators/src/diskpool/main.rs (hunk 33 above)
Comment (abhilashshetty04) [reply k8s/operators/src/diskpool/main.rs]: @tiagolobocastro We will tarnsition to Terminating now on all states. Made changes

**Review by tiagolobocastro — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors merge

File: k8s/operators/src/diskpool/main.rs (hunk 33 above)
Comment (blaisedias) [reply k8s/operators/src/diskpool/main.rs]: It is indeed covered by the BDD- create and a delete on a k8s node not provisioned with the IO engine.
